Mathlib/NumberTheory/FLT/Three.lean

```diff
@@ -342,6 +364,292 @@ lemma exists_Solution_of_Solution' : ∃ (S₁ : Solution hζ), S₁.multiplicit
+    associated_of_dvd_a_add_eta_mul_b_of_dvd_a_add_eta_sq_mul_b S hp ?_ ?_]
+  · exact p_dvd_y
+  · rw [y_spec]
+    exact dvd_mul_of_dvd_right p_dvd_y (η - 1)
+  · convert dvd_mul_of_dvd_right p_dvd_z (η - 1) using 1
+    rw [z_spec, coe_eta]
+
+private lemma x_mul_y_mul_z_eq_u_mul_w_cube : S.x * S.y * S.z = S.u * S.w ^ 3 := by
+  suffices hh : λ ^ (3 * S.multiplicity - 2) * S.x * λ * S.y * λ * S.z =
+      S.u * λ ^ (3 * S.multiplicity) * S.w ^ 3 by
+    rw [show λ ^ (3 * multiplicity S - 2) * x S * λ * y S * λ * z S =
+      λ ^ (3 * multiplicity S - 2) * λ * λ * x S * y S * z S by ring] at hh
+    have := S.two_le_multiplicity
+    rw [mul_comm _ (λ ^ (3 * multiplicity S)), ← pow_succ, ← pow_succ,
+      show 3 * multiplicity S - 2 + 1 + 1 = 3 * multiplicity S by omega, mul_assoc, mul_assoc,
+      mul_assoc] at hh
+    simp only [mul_eq_mul_left_iff, pow_eq_zero_iff', hζ.zeta_sub_one_prime'.ne_zero, ne_eq,
+      mul_eq_zero, OfNat.ofNat_ne_zero, false_or, false_and, or_false] at hh
+    convert hh using 1
+    ring
+  simp only [← x_spec, mul_assoc, ← y_spec, ← z_spec]
+  simp only [mul_comm 3, pow_mul, ← mul_pow, ← w_spec]
+  rw [← S.H, a_cube_add_b_cube_eq_mul]
+  ring
+
+private lemma x_eq_unit_mul_cube : ∃ (u₁ : (𝓞 K)ˣ) (X : 𝓞 K), S.x = u₁ * X ^ 3 := by
+  have h1 : S.x * (S.y * S.z * S.u⁻¹) = S.w ^ 3 := by
+    simp [← mul_assoc, x_mul_y_mul_z_eq_u_mul_w_cube, mul_comm _ (S.w ^ 3)]
+  have h2 : IsCoprime S.x (S.y * S.z * S.u⁻¹) :=
+    (isCoprime_mul_unit_right_right (Units.isUnit _) S.x _).2 <|
+      IsCoprime.mul_right S.isCoprime_x_y S.isCoprime_x_z
+  rcases exists_associated_pow_of_mul_eq_pow' h2 h1 with ⟨X, ⟨u₁, hX⟩⟩
+  exact ⟨u₁, X, by simp [← hX, mul_comm]⟩
+
+private lemma y_eq_unit_mul_cube : ∃ (u₂ : (𝓞 K)ˣ) (Y : 𝓞 K), S.y = u₂ * Y ^ 3 := by
+  have h1 : S.y * (S.x * S.z * S.u⁻¹) = S.w ^ 3 := by
+    rw [← mul_assoc, ← mul_assoc S.y, mul_comm S.y, x_mul_y_mul_z_eq_u_mul_w_cube]
+    simp only [mul_comm _ (S.w ^ 3), mul_assoc, mul_right_inv, Units.mul_inv, mul_one]
+  have h2 : IsCoprime S.y (S.x * S.z * S.u⁻¹) :=
+    (isCoprime_mul_unit_right_right (Units.isUnit _) S.y _).2 <|
+      IsCoprime.mul_right S.isCoprime_x_y.symm S.isCoprime_y_z
+  rcases exists_associated_pow_of_mul_eq_pow' h2 h1 with ⟨Y, ⟨u₂, hY⟩⟩
+  exact ⟨u₂, Y, by simp [← hY, mul_comm]⟩
+
+private lemma z_eq_unit_mul_cube : ∃ (u₃ : (𝓞 K)ˣ) (Z : 𝓞 K), S.z = u₃ * Z ^ 3 := by
+  have h1 : S.z * (S.x * S.y * S.u⁻¹) = S.w ^ 3 := by
+    rw [← mul_assoc, ← mul_assoc S.z, mul_comm S.z, mul_assoc S.x, mul_comm S.z, ← mul_assoc,
+      x_mul_y_mul_z_eq_u_mul_w_cube]
+    simp only [mul_comm _ (S.w ^ 3), mul_assoc, mul_right_inv, Units.mul_inv, mul_one]
+  have h2 : IsCoprime S.z (S.x * S.y * S.u⁻¹) :=
+    (isCoprime_mul_unit_right_right (Units.isUnit _) S.z _).2 <|
+      IsCoprime.mul_right S.isCoprime_x_z.symm S.isCoprime_y_z.symm
+  rcases exists_associated_pow_of_mul_eq_pow' h2 h1 with ⟨Z, ⟨u₃, hZ⟩⟩
+  exact ⟨u₃, Z, by simp [← hZ, mul_comm]⟩
+
```

Comment:
I think it would make sense to generalize [exists_associated_pow_of_mul_eq_pow'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/PrincipalIdealDomain.html#exists_associated_pow_of_mul_eq_pow') to have `Associated (a * b) (c ^ k)` as an assumption and a conjunction (`a` and `b` both are associated to a `k`th power) as the conclusion. Then this could simply be applied twice to get all three statements above.

Reply:
Done, thanks!

Reply:
What I had in mind is
```lean
variable {R} [CommRing R] [IsDomain R] [IsBezout R]

theorem exists_associated_pow_of_associated_mul_pow {a b c : R} (hab : IsCoprime a b) {k : ℕ}
    (h : Associated (a * b) (c ^ k)) : ∃ d : R, Associated a (d ^ k) := by
  obtain ⟨u, hu⟩ := h
  rw [mul_assoc] at hu
  obtain ⟨x, hx⟩ := exists_associated_pow_of_mul_eq_pow'
    ((isCoprime_mul_unit_right_right u.isUnit a b).mpr hab) hu
  exact ⟨x, hx.symm⟩

theorem exists_associated_pow_of_associated_mul_pow' {a b c : R} (hab : IsCoprime a b) {k : ℕ}
    (h : Associated (a * b) (c ^ k)) :
    (∃ x : R, Associated a (x ^ k)) ∧ ∃ y : R, Associated b (y ^ k) :=
  ⟨exists_associated_pow_of_associated_mul_pow hab h,
    exists_associated_pow_of_associated_mul_pow hab.symm (mul_comm a b ▸ h)⟩
```
and then
```lean
private lemma exists_cube_associated : 
    (∃ (X : 𝓞 K), Associated S.x (X ^ 3)) ∧ (∃ (Y : 𝓞 K), Associated S.y (Y ^ 3) ) ∧
      ∃ (Z : 𝓞 K), Associated S.z (Z ^ 3) := by
  obtain ⟨⟨xy, hxy⟩, hz⟩ :=  exists_associated_pow_of_associated_mul_pow' ... -- for (S.x * S.y) * S.z ~ S.w ^ 3
  obtain ⟨hx, hy⟩ := exists_associated_pow_of_associated_mul_pow' S.isCoprime_x_y hxy
  exact ⟨hx, hy, hz⟩
```
where the arguments to the first `exists_associated_pow_of_associated_mul_pow` need to be filled in (something like `IsCoprime.mul_left S.isCoprime_x_z S.isCoprime_y_z` and the statement `Associated (S.x * S.y * S.z) (S.w ^ 3)`, which is still proved three times in the current code (in slight variations)).
(See [here](https://live.lean-lang.org/#code=import%20Mathlib%0A%0Avariable%20%7BR%7D%20%5BCommRing%20R%5D%20%5BIsDomain%20R%5D%20%5BIsBezout%20R%5D%0A%0Atheorem%20exists_associated_pow_of_associated_mul_pow%20%7Ba%20b%20c%20%3A%20R%7D%20(hab%20%3A%20IsCoprime%20a%20b)%20%7Bk%20%3A%20%E2%84%95%7D%0A%20%20%20%20(h%20%3A%20Associated%20(a%20*%20b)%20(c%20%5E%20k))%20%3A%20%E2%88%83%20d%20%3A%20R%2C%20Associated%20a%20(d%20%5E%20k)%20%3A%3D%20by%0A%20%20obtain%20%E2%9F%A8u%2C%20hu%E2%9F%A9%20%3A%3D%20h%0A%20%20rw%20%5Bmul_assoc%5D%20at%20hu%0A%20%20obtain%20%E2%9F%A8x%2C%20hx%E2%9F%A9%20%3A%3D%20exists_associated_pow_of_mul_eq_pow'%0A%20%20%20%20((isCoprime_mul_unit_right_right%20u.isUnit%20a%20b).mpr%20hab)%20hu%0A%20%20exact%20%E2%9F%A8x%2C%20hx.symm%E2%9F%A9%0A%0Atheorem%20exists_associated_pow_of_associated_mul_pow'%20%7Ba%20b%20c%20%3A%20R%7D%20(hab%20%3A%20IsCoprime%20a%20b)%20%7Bk%20%3A%20%E2%84%95%7D%0A%20%20%20%20(h%20%3A%20Associated%20(a%20*%20b)%20(c%20%5E%20k))%20%3A%0A%20%20%20%20(%E2%88%83%20x%20%3A%20R%2C%20Associated%20a%20(x%20%5E%20k))%20%E2%88%A7%20%E2%88%83%20y%20%3A%20R%2C%20Associated%20b%20(y%20%5E%20k)%20%3A%3D%0A%20%20%E2%9F%A8exists_associated_pow_of_associated_mul_pow%20hab%20h%2C%0A%20%20%20%20exists_associated_pow_of_associated_mul_pow%20hab.symm%20(mul_comm%20a%20b%20%E2%96%B8%20h)%E2%9F%A9%0A%20%20%0A%0Avariable%20%7Bw%20x%20y%20z%20%3A%20R%7D%20(H%20%3A%20Associated%20(x%20*%20y%20*%20z)%20(w%20%5E%203))%0A%0Avariable%20(Hxy%20%3A%20IsCoprime%20x%20y)%20(Hxz%20%3A%20IsCoprime%20x%20z)%20(Hyz%20%3A%20IsCoprime%20y%20z)%0A%0Alemma%20exists_cube_associated%20%3A%20%0A%20%20%20%20(%E2%88%83%20X%2C%20Associated%20x%20(X%20%5E%203))%20%E2%88%A7%20(%E2%88%83%20Y%2C%20Associated%20y%20(Y%20%5E%203))%20%E2%88%A7%0A%20%20%20%20%20%20%E2%88%83%20Z%2C%20Associated%20z%20(Z%20%5E%203)%20%3A%3D%20by%0A%20%20obtain%20%E2%9F%A8%E2%9F%A8xy%2C%20hxy%E2%9F%A9%2C%20hz%E2%9F%A9%20%3A%3D%0A%20%20%20%20exists_associated_pow_of_associated_mul_pow'%20(IsCoprime.mul_left%20Hxz%20Hyz)%20H%0A%20%20obtain%20%E2%9F%A8hx%2C%20hy%E2%9F%A9%20%3A%3D%20exists_associated_pow_of_associated_mul_pow'%20Hxy%20hxy%0A%20%20exact%20%E2%9F%A8hx%2C%20hy%2C%20hz%E2%9F%A9%0A%0A).)

Reply:
(The names may need some permuting to fit with the changed order in `Associated` in the statements.)

Reply:
Let me think a little bit about it, but:
* I think the conclusion of `exists_associated_pow_of_associated_mul_pow` should be `Associated (d ^ k) a` (and not the other way) to match the one of `exists_associated_pow_of_mul_eq_pow'`. But indeed the assumption should be `Associated (c ^ k) (a * b)` to match the order.
* I am not completely sure it is worth to add `exists_associated_pow_of_associated_mul_pow'`: it is basically a twe copies of the umprimed version, and if needed can be inlined.

Reply:
I've proved your `exists_cube_associated` essentially as you proposed, but without adding `exists_associated_pow_of_associated_mul_pow'`. Thanks for the suggestion!